.travis.yml

```diff
@@ -1,37 +1,28 @@
 language: cpp
-sudo: false
-dist: trusty
-compiler:
-- g++
-os:
-- linux
+os: linux
+compiler: gcc
 
 addons:
-    apt:
-        sources:
-            - ubuntu-toolchain-r-test
-        packages:
-            - lcov
+  apt:
+    packages: lcov
 
-branches:
-  only:
-  - master
+before_script:
+- mkdir -p build && cd build
 
-script: ./run_build.sh
+script:
+- cmake -DCODE_COVERAGE=ON ..
+- cmake --build . --config Debug -- -j $(nproc)
+- ctest -j $(nproc) --output-on-failure
 
 after_success:
-    # Creating report
-  - cd ${TRAVIS_BUILD_DIR}
-  - lcov --directory . --capture --output-file coverage.info # capture coverage info
-  - lcov --remove coverage.info '/usr/*' --output-file coverage.info # filter out system
-  - lcov --list coverage.info #debug info
-  # Uploading report to CodeCov
-  - bash <(curl -s https://codecov.io/bash) || echo "Codecov did not collect coverage reports"
+# Create additional lcov report
+- cd ${TRAVIS_BUILD_DIR}
```

Comment:
This is wrong. You cd into `build` before, here you cd into the root dir but call `lcov` with `--directory .` I suggest:

- Combine before_script and script and cd out of the build dir at the ende of script. Add a note there to put this into a bash file with `set -e` (this avoids problems like e.g. cmake fails but make is run which is an annoying travis issue)
- remove this cd and use `lcov --directory build`

Reply:
I cannot find a reference stating lcov needs to run in the build directory. Setting the folder to `build` would reduce the search space a bit.
I'll remove the second `cd`. That makes the `lcov` commands cleaner too.

I've moved the directory creation to the `before_script:` because I found it made the `script:` crowded and harder to read. I wanted the focus on the `cmake` and `ctest` commands.

I like your reminder of `set -e`, but I'm not sure it belongs in this example.
Aside: You know it does not require the use of a separate bash file? I did some experimenting lately. Make it the first (before_)script command and it works for the rest of the .travis.yml file.

Reply:
> I cannot find a reference stating lcov needs to run in the build directory. Setting the folder to build would reduce the search space a bit.

Correct. Not setting it wastes precious time on CI by searching unrelated folders

> I've moved the directory creation to the before_script:

IMO the sequence `mkdir build && cd build && cmake && make && make install` is common enough for a build(script). Also returning from there (or using a bash file) makes it easier to reason about where you are in subsequent steps

> I like your reminder of set -e, but I'm not sure it belongs in this example.

As this is an example on HOW to do this IMO this belongs here. As mentioned the problem is that travis runs the make commands even when the cmake command failed! This is a bug and can cause all sorts of problems depending what you do afterwards (maybe uploads?)

> Aside: You know it does not require the use of a separate bash file?

It actually does. If you set it in the travis file you will confuse the build bot and it will swallow your error: https://travis-ci.community/t/build-fails-with-no-obvious-reason/1452/4

Reply:
The focus I want is not on a known sequence, but on:
- running ctest after the build,
- `-DCODECOVERAGE=ON` (which corresponds to the option in the CMakeLists.txt). Because it creates a hook into CMake for Travis job customisation.

I think limiting the .travis.yml to a single 'change directory' should help in reasoning.
Not a good practice for more complicated scripts, but I believe this example serves to show the basics, such that they are easy to adapt to other projects. (With their own styles for scripting and directory management.)

Thanks for the info on the `set -e` issue.  
I like the idea of giving more attention to its use, but feel it is out of scope for this example.
And with the issues being considered by Travis I do not feel qualified to suggest a best practice.
Might I suggest making a PR when you have a good implementation? And maybe adding to the TravisCI documentation?

Reply:
I agree mostly. My comment was based on me being unsure if travis resets the working directory between different steps. I'm pretty sure it doesn't but there is always some doubt that bothers me.

> Not a good practice for more complicated scripts

My point is that people might start with this code and then add their own stuff around it without conceptually changing it. So using "bad practices" here might lead to people keeping that. :/

> And with the issues being considered by Travis I do not feel qualified to suggest a best practice.

Travis specifically says to NOT use it in .travis.yml but only in scripts.